src/systems/filecoin_blockchain/storage_power_consensus/storage_power_consensus_subsystem.go

```diff
@@ -57,12 +52,15 @@ func (spc *StoragePowerConsensusSubsystem_I) ValidateBlock(block block.Block_I)
 	// return nil
 }
 
-func (spc *StoragePowerConsensusSubsystem_I) validateTicket(ticket block.Ticket, pk filcrypto.PublicKey) bool {
-	panic("")
-	// Randomness1 := storagePowerConsensus.GetTicketProductionSeed(sms.CurrentChain, sms.Blockchain.LatestEpoch())
-	// input := VRFPersonalizationTicket
-	// input.append(Randomness1)
-	// return ticket.Verify(input, pk)
+func (spc *StoragePowerConsensusSubsystem_I) validateTicket(ticket block.Ticket, pk filcrypto.VRFPublicKey, minerAddr addr.Address) bool {
+	randomness1 := spc.GetTicketProductionSeed(spc.blockchain().BestChain(), spc.blockchain().LatestEpoch())
+
+	var input []byte
+	input = append(input, byte(filcrypto.DomainSeparationTag_Case_Ticket))
+	input = append(input, randomness1...)
+	input = append(input, byte(filcrypto.InputDelimeter_Case_Bytes))
+	input = append(input, minerAddr.ToBytes()...)
```

Comment:
Hmm... I don't think it's feasible in general to duplicate serialization code for each type involved in randomness generation -- this will tend to cause lots of subtle errors and be very difficult to maintain over time. I've drafted a candidate API that I think solves this in a generic way, and also leverages the existing serialization machinery to avoid potential redundancy there:

https://github.com/filecoin-project/specs/pull/679

Let me know what you think -- if it looks like it will work for these use cases, maybe easiest to rebase (parts of?) this PR on that branch and swap in the new API.

Reply:
Happy to use your API which is nice though what is the index for in it?

Also I am not sure what you mean wrt my duplicating serialization code. What I've implemented is most of it, as in 
- how to generate new tickets
- how to draw tickets
- how to generate post randomness.

I might be able to generalize the methods a bit (this is where your API is nice) by passing in the tag and serialization... But then some of the types do need to be different (e.g. generation vs sampling).

If you can give me a bit more to go on I'll rebase and try again

Reply:
> Happy to use your API which is nice though what is the index for in it?

The index is intended to be used for generation of strings longer than 256 bits (e.g., a 768-bit pseudorandom string may be derived as: `tag.DeriveRandWithIndex(s, 1) || tag.DeriveRandWithIndex(s, 2) || tag.DeriveRandWithIndex(s, 3)` where || indicates concatenation.) I'll update #679 with a comment to this effect.

Reply:
> Also I am not sure what you mean wrt my duplicating serialization code. 

I mean writing multiple implementations of transforming input objects to pseudorandom output (which might vary in its details based on the type of the input object, but still duplicates the same basic logic in each case).

> I might be able to generalize the methods a bit (this is where your API is nice) by passing in the tag and serialization... But then some of the types do need to be different (e.g. generation vs sampling).

I think generation should be covered by the API already present in #679. What functionality are you looking for in terms of sampling? Most likely we'll be able to implement this also in a generic way in `randomness.go`.